File: luci-app-mqttwol/luasrc/model/cbi/mqttwol.lua
--[[
CBI mapping for /etc/config/mqttwol (named section main).
]]

local sys = require "luci.sys"

local m = Map(
	"mqttwol",
	translate("MQTT Wake-on-LAN"),
	translate(
		"Runs mosquitto_sub in the background via procd and calls etherwake for each MQTT payload matching a MAC address."
	)
)
m:chain("services")

local s = m:section(NamedSection, "main", "mqttwol", translate("Settings"))
s.addremove = false
s.optional = false

local en = s:option(Flag, "enabled", translate("Enable"))
en.rmempty = false

local srv = s:option(Value, "server", translate("MQTT broker address"))
srv.datatype = "host"
srv.rmempty = false

local pt = s:option(Value, "port", translate("MQTT broker port"))
pt.placeholder = "1883"
pt.datatype = "port"
pt.rmempty = false

local usr = s:option(Value, "username", translate("MQTT username"))
usr.optional = true
usr.placeholder = translate("leave empty when broker allows anonymous subscriptions")

local pw = s:option(Value, "password", translate("MQTT password"))
pw.password = true
pw.optional = true
pw.rmempty = true

local tpic = s:option(Value, "topic", translate("MQTT topic"))
tpic.datatype = "string"
tpic.rmempty = false
tpic.placeholder = "home/router/wol"

local ifname = s:option(Value, "interface", translate("Ethernet interface"))
ifname.datatype = "string"
ifname.rmempty = false
ifname.placeholder = "br-lan"
ifname.description = translate(
	"Kernel interface name passed to etherwake -i; br-lan is the LAN bridge used by default OpenWrt images."
)

m.on_after_commit = function(_)
	sys.call("/etc/init.d/mqttwol restart >/dev/null 2>&1")
end

return m

File: luci-app-mqttwol/htdocs/luci-static/resources/view/mqttwol.js
'use strict';
'require view';
'require form';
'require uci';
'require fs';
'require tools.widgets as widgets';

return view.extend({
	load: function() {
		return uci.load('mqttwol');
	},

	render: function() {
		const m = new form.Map(
			'mqttwol',
			_('MQTT Wake-on-LAN'),
			_('MQTT subscription and Wake-on-LAN via mosquitto_sub and etherwake.')
		);

		const s = m.section(form.NamedSection, 'main', 'mqttwol', _('Settings'));

		let o;

		o = s.option(form.Flag, 'enabled', _('Enable'));

		o = s.option(form.Value, 'server', _('MQTT broker address'));
		o.rmempty = false;
		o.datatype = 'host';

		o = s.option(form.Value, 'port', _('MQTT broker port'));
		o.placeholder = '1883';
		o.datatype = 'port';
		o.rmempty = false;

		o = s.option(form.Flag, 'use_auth', _('MQTT use authentication'));
		o.default = o.disabled;

		o = s.option(form.Value, 'username', _('MQTT username'));
		o.depends('use_auth', '1');
		o.datatype = 'string';
		o.optional = true;

		o = s.option(form.Value, 'password', _('MQTT password'));
		o.password = true;
		o.depends('use_auth', '1');
		o.datatype = 'string';
		o.optional = true;

		o = s.option(form.Value, 'topic', _('MQTT topic'));
		o.rmempty = false;
		o.datatype = 'string';

		o = s.option(
			widgets.DeviceSelect,
			'interface',
			_('Ethernet interfaces'),
			_('Network devices for etherwake (same control as on the interfaces page). Choose one or more; each gets a magic packet.')
		);
		o.multiple = true;
		o.rmempty = false;

		const origSave = m.save.bind(m);
		m.save = function(cb, silent) {
			function beforeSave(/* ... */) {
				if (uci.get('mqttwol', 'main', 'use_auth') != '1') {
					uci.unset('mqttwol', 'main', 'username');
					uci.unset('mqttwol', 'main', 'password');
				}
				const args = arguments;
				return (typeof cb == 'function')
					? cb.apply(null, args)
					: undefined;
			}

			return origSave(beforeSave, silent).then(function() {
				return fs.exec_direct('/etc/init.d/mqttwol', [ 'restart' ]);
			});
		};

		return m.render();
	},
});
